### PR TITLE
WT-11603 Proof-of-concept implementation of direct packing functions with macros

### PR DESCRIPTION
**Pull request by ershov**

### CODE REVIEW

**Comment by ershov** [comment]
Hey this is a proof-of-concept of a macros-based implementation. Some advantages are:
* No externally generated code - pure C implementation.
* Functions are created on-demand.

Let me know what you think. If we decide to go with this one, I'll tidy it up and thoroughly test it before sending a final PR.

**Comment by kommiharibabu** [comment]
@ershov I like this C macro approach; it will be simple to add a new record type to get the auto-generated packing and unpacking logic.

File: src/include/packing_inline.h [hunk 1]
```diff
@@ -754,3 +754,294 @@ __wt_struct_size_adjust(WT_SESSION_IMPL *session, size_t *sizep)
...
+
+
+/*
+ * Specialized packing functions implementations
+ */
+
+WT_DEFINE_PACKING(system_record,  /* __wt_size_system_record(), __wt_pack_system_record(), __wt_unpack_system_record() */
+    uint32_t, rectype)
+
+WT_DEFINE_PACKING(checkpoint_start,  /* __wt_size_checkpoint_start(), __wt_pack_checkpoint_start(), __wt_unpack_checkpoint_start() */
```
Comment (agorrod) [inline src/include/packing_inline.h:1009]: I haven't dug deeply into what this replaces, but I liked that the prior version just required a change to "dist/log_data.py" (which allows for records to be described in a natural way), then the relevant code was generated and correct.

It is nice not having to manually edit multiple locations when adding log record types, especially when it is done so rarely.

Comment (ershov) [reply src/include/packing_inline.h:1009]: Yeah I have to agree. Variadic macros don't work with older MSVC and it doesn't fit into multi-inclusion pattern.

I'll revert back to the Python generation.

**Comment by ershov** [comment]
> @ershov, in terms of C code executed (and thus performance), is this any different than the PR that @kommiharibabu did?
> 
> It's pretty amazing to see the macro machinery involved here. We'd need some good tests to be sure that the preprocessors are up to the task.

Thanks @ddanderson!

In terms of performance they are exactly the same since the macros expand into the same code as generated by Python (I mean, that's the idea - both PRs are non-functional prototypes).

Yeah, the machinery is pretty amazing, it's sad we can't use it because MSVC 2017 has it's own (incompatible) ideas of how macro expansion works and that's the version that we use in Evergreen. It's only fixed in MSVC 2019.

**Pull request closed without merging**